gamedata/ddb.py

```diff
@@ -0,0 +1,417 @@
+        :type user_id: int
+        :type entity_type: str
+        :rtype: set[int] or None
+        """
+        log.debug(f"Getting DDB entitlements for Discord ID {user_id}")
+        user_e10s = await self._get_user_entitlements(ctx, user_id)
+        if user_e10s is None:
+            return None
+
+        entity_e10s = await self._get_entity_entitlements(ctx, entity_type)
+
+        # calculate visible entities
+        accessible = set()
+        user_licenses = user_e10s.licenses
+        for entity in entity_e10s:
+            if entity.is_free or user_licenses & entity.license_ids:
+                accessible.add(entity.entity_id)
+
+        log.debug(f"Discord user {user_id} can see {entity_type}s {accessible}")
+
+        return accessible
+
+    async def get_ddb_user(self, ctx, user_id):
+        """
+        Gets a Discord user's DDB user, communicating with the Auth Service if necessary.
+        Returns None if the user has no DDB link.
+
+        :type ctx: discord.ext.commands.Context
+        :type user_id: int
+        :rtype: BeyondUser or None
+        """
+        log.debug(f"Getting DDB user for Discord ID {user_id}")
+        user_cache_key = f"beyond.user.{user_id}"
+        unlinked_sentinel = {"unlinked": True}
+
+        cached_user = await ctx.bot.rdb.jget(user_cache_key)
+        if cached_user == unlinked_sentinel:
+            return None
+        elif cached_user is not None:
+            return BeyondUser.from_dict(cached_user)
+
+        user_claim = self._jwt_for_user(user_id)
+        token, ttl = await self._fetch_token(user_claim)
+
+        # cache unlinked if user is unlinked
+        if token is None:
+            await ctx.bot.rdb.jsetex(user_cache_key, unlinked_sentinel, USER_ENTITLEMENT_TTL)
+            return None
+
+        user = self._parse_jwt(token)
+        await ctx.bot.rdb.jsetex(user_cache_key, user.to_dict(), ttl)
+        await Stats.count_ddb_link(ctx, user_id, user)
+        return user
+
+    async def _get_user_entitlements(self, ctx, user_id):
+        """
+        Gets a user's entitlements in the current context, from cache or by communicating with DDB.
+
+        Returns None if the user has no DDB connection.
+
+        :type ctx: discord.ext.commands.Context
+        :param user_id: The Discord user ID.
+        :type user_id: int
+        :rtype: UserEntitlements
+        """
+        user_entitlement_cache_key = f"entitlements.user.{user_id}"
+        cached_user_entitlements = await ctx.bot.rdb.jget(user_entitlement_cache_key)
+        if cached_user_entitlements is not None:
+            return UserEntitlements.from_dict(cached_user_entitlements)
+
+        user = await self.get_ddb_user(ctx, user_id)
+
+        if user is None:
+            return None
+
+        # feature flag: is this user allowed to use entitlements?
+        enabled_ff = await ctx.bot.ldclient.variation("entitlements-enabled", user.to_ld_dict(), False)
```

Comment:
if the user has linked DDB but the entitlements feature flag is off, don't hit entitlements